Issue #89: Require env vars for production resources, remove app_config.json fallbacks

N1kunj1998 opened the pull request.

mamuqsit committed:
fastmcp: pin version to 2.14.0 (#98)

fastmcp.Client in v2.14.1 does not work - it hangs indefinitely

mamuqsit committed:
ci: rename staging branch (#97)

* ci: rename staging branch

* docs/deployment.md: update branch name

N1kunj1998 committed:
feat: Add WhatsApp template messages for onboarding, leaderboard, and DYK with multi-language support (#95)

* background-jobs: deprecate POST /schedule endpoint (#93)

* background-jobs: deprecate POST /schedule endpoint

* Return 200 OK with JSON "detail" as per standards
Addresses few points made by @a4i-architect

* POST /asha_logs: tweak setup, add integration tests (#92)

* POST /asha_logs: add integration tests
test_admin.py is expected to fail at two points

* admin.py: fix POST /asha_logs not returning expected response
This change is expected to fix the failing integration tests

* test_admin: use datetime.isoformat() in place of custom date formatter

* /asha_logs: stream logs server-side, process file client-side
Unlike XLSX, CSV is one of the better formats to use here as we can iteratively deliver data to user.

XLSX is a ZIP of XML files so it cannot be streamed.

* /get_bot_messages: add a 'limit=1000' parameter
We can leverage this to get a sample of messages for later /asha_logs integration tests

* test_admin: add test case to assert /asha_logs csv output format
This test will ensure the format contains few necessary attributes and that all records are sorted by incoming timestamp attribute.
/get_bot_messages is used to fetch a sample of up to a 100 messages.

* cleanup code, diff

* workflows: temporarily run cd
This change will be reverted after failing test is fixed

* Revert "workflows: temporarily run cd"

This reverts commit dbff763d4879516d839948bf142c17351613fa09.

* changed leaderboard and send dyk code to use template instade of sending normal whatsapp message

* added template message for user to get message when he is onboarded via backed script

* final commit also checked by running background jobs and made changes to now work with template code

* Quote all DYK CSV fields to avoid comma parsing errors

* Cleanup DYK templates and enforce CSV validation

---------

Co-authored-by: Muqsit <notmuqsit+a4i@gmail.com>

mamuqsit committed:
tests: cleanup, widen onboarding test to cover all languages (#100)

* refactor: cleanup onboarding test, extend to all four supported languages

* test-onboarding: validate new user attributes

* test-onboarding: cleanup requests

* test-onboarding: reformat code

* .gitignore: add coverage.json

* byoeb-integrations: cleanup tests - avoid creating persistent .wav files

* tests: patch for sqlite v3.35

N1kunj1998 committed:
Automated Monthly ASHABot Log Analysis (#82)

* inital commit for this task, will remove background jobs change in future commit

* removed background jobs part and removed unnecessary code and created new script for extracting relavent data

* reuse asha_logs.fetch_daily_logs in compile_monthly_logs script

* added code for IDK analysis

* Added more data points and results of IDK analysis in output markdown file

* background-jobs: deprecate POST /schedule endpoint (#93)

* background-jobs: deprecate POST /schedule endpoint

* Return 200 OK with JSON "detail" as per standards
Addresses few points made by @a4i-architect

* added LLM generated insights and summary

* POST /asha_logs: tweak setup, add integration tests (#92)

* POST /asha_logs: add integration tests
test_admin.py is expected to fail at two points

* admin.py: fix POST /asha_logs not returning expected response
This change is expected to fix the failing integration tests

* test_admin: use datetime.isoformat() in place of custom date formatter

* /asha_logs: stream logs server-side, process file client-side
Unlike XLSX, CSV is one of the better formats to use here as we can iteratively deliver data to user.

XLSX is a ZIP of XML files so it cannot be streamed.

* /get_bot_messages: add a 'limit=1000' parameter
We can leverage this to get a sample of messages for later /asha_logs integration tests

* test_admin: add test case to assert /asha_logs csv output format
This test will ensure the format contains few necessary attributes and that all records are sorted by incoming timestamp attribute.
/get_bot_messages is used to fetch a sample of up to a 100 messages.

* cleanup code, diff

* workflows: temporarily run cd
This change will be reverted after failing test is fixed

* Revert "workflows: temporarily run cd"

This reverts commit dbff763d4879516d839948bf142c17351613fa09.

* added charts and Key Metrics table and also code for MoM analysis

* Completed code with adding upload to Azure functionality

* added user related analysis and MoM growth and changed upload container in azure storage

* Use: Handle monthly log fetch from async generator before building DataFrame

---------

Co-authored-by: Muqsit <notmuqsit+a4i@gmail.com>

mamuqsit committed:
pydantic: make optional fields use explicit default=None (#104)

* pydantic: make optional fields use explicit default=None

According to https://docs.pydantic.dev/2.7/concepts/fields, the documented way to declare defaults is by using `default=`. This also solves VSC static analysis from underlining all such models where we expect defaults.
Furthermore:
> If you use typing.Optional, it doesn't mean that the field has a default value of None!

* message-response: make WhatsAppResponse.response_status optional in favor of test

mamuqsit committed:
kb-app: generate related questions in mr, te (#107)

* kb-app: generate related questions in mr, te as well

* /vector/search: expose related questions
This is needed to debug whether related questions are indeed being generated for all languages, and in the right script (Devanagari/Telugu)

mamuqsit committed:
has_interactive_list_additional_info: ensure list is not empty (#106)

mamuqsit committed:
bot-config: deduplicate answer_prompts, remove related_questions from prompt (#112)

mamuqsit committed:
translation: use gpt-4o-transcribe for speech-to-text (#99)

* refactor: rename async_azure_openai_whisper -> async_azure_openai_translator
This impl is generic for all OpenAI transcriber models, we can reuse it for GPT-4o Transcribe.

* mcp: support audio input in asha_chat

* dep-setup: suffix speech translators by their utility (_tts, _stt)

* app-config: allow specifying per-language stt and tts translation services

* app-config: use gpt-4o-transcribe for en, hi, te; whisper for mr
Goal was to use whisper for te as well, but whisper bails on processing te (unsupported)

* AzureOpenAIWhisperParamsEnum -> AzureOpenAISpeechParamsEnum

* dep-setup: make speech_* require user

* stt: gate gpt-4o-transcribe behind feature flag 'stt_latency_mitigation'

* tests: add integration tests for en, hi audio inputs

* azure-speech: fix failing test

* tests/integrations: add sample ogg files for audio messages

* AZURE_OPENAI_SPEECH_ENDPOINT: bail when undefined

* speech: reuse models having identical configuration

* speech_* selectors: accept lang, test_user instead of bulky User object for matching

* app-config: update speech model api versions 2024-06-01 -> 2024-10-21

* refactor: introduce a singular translator adapter for stt and tts

* fix failing test case
We shouldn't be using dependency_setup.py in our unit tests, but this is sufficient for time being

* dep-setup: read azure cognitive resource ids from env

mamuqsit committed:
mongodb: retire abstraction layers (#81)

* mongodb: first look at replacing abstractions with direct invocation

* mongodb: use mongomock library for unit tests

* byoeb-integrations: drop 'database' namespace

* repos: fix inconsistent typing

* ci/cd: use same staging-deploy environment as azure webapp (#83)

* ci-cd: temporarily enlist muqsit/docker-azure-env as staging branch

* Dockerfile: use python:3.11-slim-bullseye
This commit is expected to cause a deployment failure

* chat-app: overwrite default sqlite3 library with pysqlite3-binary
Expecting a successful deployment...

* Revert "ci-cd: temporarily enlist muqsit/docker-azure-env as staging branch"

This reverts commit 4fb32fb6fb353d4b6cad18737266f9d7421a02ad.

* Prompt modifications (#80)

* openapi: further documentation enhancements (#79)

* /get_users: use query params in place of request body
GET requests cannot have a request body according to HTTP specs - curl allows a GET with body anyway, but JavaScript's fetch() bails to process these requests.

* ashabot: model a PhoneNumberId type for auto validation

* apis/user: cleanup examples
These examples will be autogenerated by FastAPI based on pydantic models

* router: remove redundant 'response_class' field
These should be derived from the function return value itself

* apis/background_jobs: code cleanup

* apis/chat: code cleanup

* /get_users: convert to POST

* user: add integration tests for api endpoints

* scripts/onboarding.py: code cleanup

* tests: add integration test for onboarding.py

* apis: specify tags in router instead of in individual endpoints

* apis: code cleanup

* byoeb-core: add unit test for phone number validation

* byoeb-core: add coverage opts for pytest

* setup_scheduled_jobs: fail fast on exceptions
Any errors in job configuration will now cause deployment to fail instead of silently ignoring setup errors.

* scripts/onboarding.py: extract users only when needed

* message-db: delete unused delete_message_collection

* Prompt modifications (#80)

* openapi: further documentation enhancements (#79)

* /get_users: use query params in place of request body
GET requests cannot have a request body according to HTTP specs - curl allows a GET with body anyway, but JavaScript's fetch() bails to process these requests.

* ashabot: model a PhoneNumberId type for auto validation

* apis/user: cleanup examples
These examples will be autogenerated by FastAPI based on pydantic models

* router: remove redundant 'response_class' field
These should be derived from the function return value itself

* apis/background_jobs: code cleanup

* apis/chat: code cleanup

* /get_users: convert to POST

* user: add integration tests for api endpoints

* scripts/onboarding.py: code cleanup

* tests: add integration test for onboarding.py

* apis: specify tags in router instead of in individual endpoints

* apis: code cleanup

* byoeb-core: add unit test for phone number validation

* byoeb-core: add coverage opts for pytest

* setup_scheduled_jobs: fail fast on exceptions
Any errors in job configuration will now cause deployment to fail instead of silently ignoring setup errors.

* scripts/onboarding.py: extract users only when needed

* message-db: delete unused delete_message_collection

* repositories: move generic methods to a mongodb base repo class

* mongodb repo: find_all - use good default values

* mongodb base repo: return a (buffered) asynciterator instead of loading all records in memory

* message repo: return a buffered AsyncIterator instead of in-memory lists

* dyk: use same buffered fetch strategy

* ashabot: update cases to utilize new buffered fetch strategy

* repo: fix class hierarchy

* background-jobs: deprecate POST /schedule endpoint (#93)

* background-jobs: deprecate POST /schedule endpoint

* Return 200 OK with JSON "detail" as per standards
Addresses few points made by @a4i-architect

* POST /asha_logs: tweak setup, add integration tests (#92)

* POST /asha_logs: add integration tests
test_admin.py is expected to fail at two points

* admin.py: fix POST /asha_logs not returning expected response
This change is expected to fix the failing integration tests

* test_admin: use datetime.isoformat() in place of custom date formatter

* /asha_logs: stream logs server-side, process file client-side
Unlike XLSX, CSV is one of the better formats to use here as we can iteratively deliver data to user.

XLSX is a ZIP of XML files so it cannot be streamed.

* /get_bot_messages: add a 'limit=1000' parameter
We can leverage this to get a sample of messages for later /asha_logs integration tests

* test_admin: add test case to assert /asha_logs csv output format
This test will ensure the format contains few necessary attributes and that all records are sorted by incoming timestamp attribute.
/get_bot_messages is used to fetch a sample of up to a 100 messages.

* cleanup code, diff

* workflows: temporarily run cd
This change will be reverted after failing test is fixed

* Revert "workflows: temporarily run cd"

This reverts commit dbff763d4879516d839948bf142c17351613fa09.

* repositories: delete example usage file
This ought to be placed outside source, preferably in a docs/dev.md or the repo wiki

* repositories: cleanup comments
Docstrings already exist in base classes, so impls need not redefine them

* test-leaderboard: use mongomock-motor to clean up a plethora of monkeypatching

---------

Co-authored-by: Soumabha Ray Chaudhuri <Soumav@gmail.com>

N1kunj1998 committed:
corrected truncating text (#109)

N1kunj1998 committed:
Add unit tests to filter Test District from leaderboard before WhatsApp API calls (#111)

mamuqsit committed:
messenger ui: first look at frictionless chat testing (#103)

* asha_chat: reuse existing user flows for chat handling

* /chat: implement a messenger ui

* /chat: gate behind env ASHABOT_UAT=1

* cleanup: remove unused dep on ChannelClientFactory

* test-onboarding: simplify by using new dummy channel

* /chat: improve responsiveness of ui

* /chat: introduce support for audio outputs

* mcp: fix asha_chat returning question rather than the answer

* ui: make header and input field sticky

* ui: add 'Loading messages' feedback

* ...

* ui: add support for audio input

* /get_bot_messages: cleanup naming and doc

* ui: display related questions

* ui: allow one audio to play at a time

* ui: be more specific on what the delete button does

* ui: add favicon to avoid /favicon.ico requests

* ui: optimize /get_bot_messages by tracking timestamp

* ui: fix longer audios not being b64 encoded correctly

* ui: display time taken to generate message

* ui: fix old messages apparing after relogging

* chat: add disambiguation categories (future)

N1kunj1998 committed:
feat: Migrate to async-only APIs across codebase (#110)

* Add async-only migration validation test suite

Validation: run `pytest test_async_migration_validation.py -v` (or phase-specific classes) to confirm migration changes are correct.

* Make core base classes async-only

- BaseLLM: drop sync generate_response; rename agenerate_response → generate_response (async)
- BaseVectorStore: remove sync add/update/delete/retrieve; rename async methods (remove “a”), include retrieve_similar_chunks
- BaseChannel: remove sync send/receive; rename async methods (remove “a”), fix receive typo; keep interactive send async-only
- BaseQueue: remove sync send/receive/delete; rename async methods (remove “a”)

Validation:
- From repo root: pytest test_async_migration_validation.py::TestPhase1_BaseLLM \
    test_async_migration_validation.py::TestPhase1_BaseVectorStore \
    test_async_migration_validation.py::TestPhase1_BaseChannel \
    test_async_migration_validation.py::TestPhase1_BaseQueue -v

* Migrate implementations to async-only APIs (Phase 2)

Update all concrete implementations to match async-only base class APIs:
- LLM implementations: Remove sync methods, rename agenerate_response → generate_response
- Vector Store implementations: Remove sync methods, rename async methods (remove 'a' prefix)
- Message Queue implementations: Remove sync methods, rename async methods (remove 'a' prefix)
- Add trio dependency for test support

All Phase 2 validation tests pass ✅

* Update call sites to async-only method names (Phase 3)

Update all call sites to use renamed async methods (removed 'a' prefix):
- agenerate_response → generate_response
- aretrieve_top_k_chunks → retrieve_top_k_chunks
- aadd_chunks → add_chunks
- asend_message/receive_message/delete_message → send_message/receive_message/delete_message

Updated 19 files across services, scripts, integrations, and tests.
All Phase 3 validation tests pass ✅

* Update test files to async-only API (Phase 4)

- Convert test functions to async with @pytest.mark.asyncio
- Fix ChromaDBVectorStore.add_chunks() to return async generator
- Update byoeb-core dependency to fix VectorStoreMetadata import

Updated 2 files: test_vector_store_chroma.py, chroma/base.py
All Phase 4 validation tests pass ✅

* Review comments addressed: removed trio dependency from pyproject.toml as it was not being used and removed test_async_migration_validation.py

N1kunj1998 committed:
feat(security): Remove production resource fallbacks, require env vars to prevent accidental production access (Issue #89)

N1kunj1998 committed:
Require environment variables for production configs, remove app_config.json placeholders

mamuqsit commented on byoeb-v1/byoeb/byoeb/chat_app/configuration/dependency_setup.py:
```diff
@@ -51,8 +51,15 @@ def wrapper(*args, **kwargs):
+        raise ValueError(
+            "APP_LOGGER_NAME environment variable must be set. "
+            "This prevents accidental access to production resources. "
+            "Set it in keys.env (staging or production section)."
```
This is overly specific to our internal reasoning for the change, we can simplify the message to just the first line:
`raise ValueError("APP_LOGGER_NAME environment variable must be set.")`

N1kunj1998 replied:
Done

mamuqsit commented on byoeb-v1/byoeb/byoeb/chat_app/app_config.json:
```diff
@@ -35,27 +36,27 @@
             "message_collection": "ashamessages",
-            "jobs_collection": "ashajobs"
+            "jobs_collection": "ashajobs",
+            "_comment": "Override MONGO_DB_DATABASE_NAME via environment variables for production"
```
`_comment` entries can be removed from the config - the new ValueErrors being thrown will make mandatory envs visible early during runtime

N1kunj1998 replied:
removed all `_comment` from app_config.json

mamuqsit commented on byoeb-v1/byoeb/byoeb/kb_app/app_config.json:
```diff
@@ -27,9 +30,10 @@
+            "account_url": "https://your-storage-account.blob.core.windows.net",
+            "container_name": "your-container",
+            "analysis_container_name": "your-analysis-container",
+            "_comment": "Override AZURE_STORAGE_BLOB_ACCOUNT_URL, AZURE_STORAGE_CONTAINER_NAME, and AZURE_STORAGE_ANALYSIS_CONTAINER_NAME via environment variables for production"
```
Placeholder values can be removed since we aren't reading these from config anymore.

N1kunj1998 replied:
Removed all Placeholder Values from app_config.json

N1kunj1998 committed:
Address PR review: simplify error messages and clean up config files

a4i-architect reviewed (commented):
### Code Review Summary

---

#### General Observations:
- The updates are mostly refinements: correcting branch references, adjusting default values (`default=None`), and softening type constraints.
- Extensive normalization of optional fields with `default=None` improves consistency.
- Several classes have duplicate or commented-out code (notably in `mongo_db/base.py`) that should be removed to avoid clutter.
- Method signatures for async methods now correctly use `async def` with `await` where needed, which is a good improvement.
- Minor typo issues: for instance, in `byoeb_core/models/whatsapp/incoming/status_message.py`, the `status` field’s default value should be 'read', which is correctly set via `default='read'`.

---

#### Specific Issues & Recommendations:

**1. `.github/workflows/ci-cd-pipeline.yml` correction**  
Good fix on branch conditions. Consider consolidating branch names into variables or using pattern matching for cleaner workflows.

**2. Type Annotations & Optional Fields**  
- Consistent use of `default=None` over `Optional[type] = None`. Instead, prefer one style.  
- In `BaseChannel`, `async def send_message()` now correctly marked; similar improvements in other async classes (message_queue, vector_stores).  
- In `message_context.py`, all optional fields are properly defaulted, improving robustness.

**3. Removed or commented code in `mongo_db/base.py`**  
Indicative of code vetting: should be cleaned permanently to avoid confusion.

**4. Async Methods & `@field_validator`**  
- Proper `@field_validator` with `mode="before"` in `audio_convertor.py` is good practice.  
- Async method signatures are now properly declared, which is crucial.

**5. Code Quality & Duplication**  
- Multiple classes have redundant code, such as in `whatsapp/incoming/status_message.py` and `message_response.py`.  
- Consider removing deprecated commented code blocks from `base.py` entirely.
- Some classes like `MediaData` in `media_request.py` are inline with all fields marked as Optional, which is verbose. Explicit defaults improve clarity.

**6. Data Models & Typing**  
- Usage of `Optional[...]` combined with `default=None` is proper, but some fields use `default=None` while also being `Optional`. This is redundant but acceptable.  
- Fields like `context` or `media` have `default=None`, which allows easier handling of partial payloads.

**7. Implementation Specifics**  
- Classes with only `abstractmethod`s are correctly marked as `ABC`.  
- Slight inconsistencies in naming conventions: e.g., `_response_status`, `status` fields are clear but consider renaming to avoid ambiguity.

**8. Potential Over-complication**  
- Interfaces like `BaseLLM`, `BaseSpeechTranslator`, `BaseQueue`, etc., are standard abstract classes.  
- Cyclomatic complexity appears low; no overly complex logic is detected from the diff.

---

#### Critical Missing Checks:
- No validation of data formats or schema validation beyond basic Pydantic types.  
- Consider adding docstrings or comments explaining purpose of each class for maintainability.  
- The `vector_stores/azure/azure_search.py` class could benefit from typing for list (e.g., `List[str]`), especially for `text_vector_3072`.

---

### Final Score: 85/100

**Rationale:**  
The updates are mostly correct, well-typed, and address previous issues. Minor cleanup (removing commented code, standardizing defaults) and stricter validation would push this higher. The structure supports clean extension.

---

### Recommendations:
- Remove commented-out or duplicate code files.
- Standardize usage of `Optional[...]` vs. `default=None`.
- Add class-level docstrings for clarity.
- For async interfaces, consider providing concrete awaitable implementations or at least placeholder comments.
- Finalize any unimplemented abstract methods (`pass`) with logging or NotImplementedError, depending on use case.

---

**In essence, the code is acceptable for acceptance with minor refinements.**

N1kunj1998 committed:
feat: Require queue names from environment variables

- Queue names now required via AZURE_QUEUE_* environment variables
- No fallback to app_config.json to prevent accidental production access
- Fail fast validation at startup if queue names not set

a4i-architect reviewed (commented):
### Overall Summary
The proposed modifications mainly involve:
- Updating branch names for workflows
- Renaming method signatures (mostly adding `async`), fixing type hints, and setting default values (`None`/`default=None`)
- Minor code standardizations and cleanup (e.g., `Optional` fields consistency, removing commented-out code)
- Extending some class abstract methods with `async`/`await` pattern, aligned with interfaces

---

### Detailed Review & Issues

#### 1. **Workflow YAML modifications**
- Corrects branch names; good.
- Conditional logic for publishing/pipeline failure notifications: consistent and correct (`a4i/staging`/`a4i/main`).

#### 2. **`.gitignore`**
- Additions (`coverage.json`) are harmless but ensure it’s intentionally needed.

#### 3. **Abstract classes (`*_base.py`)**
- **Type annotations**: The `ABC` classes should consistently mark methods as `async` if they are meant to be async.  
- **Inconsistent method signatures**: Many abstract methods switch between `def` and `async def` or omit `async` where it makes sense.  
- **Dead code / commented methods**: Some no-op or commented methods (e.g., in `BaseChannel`) could be cleaned out or properly documented.

**Recommendations:**  
- Clearly specify which methods are `async` across classes.  
- Remove commented-out or duplicated abstract method signatures to avoid confusion.

#### 4. **Type Hints & Optional fields**
- **Correct**: The usage of `Optional` with default values (`default=None`) is consistent but could be simplified to `Optional[...] = None`.  
- **In `message_context` classes**: Fields like `mime_type`, `media_type`, etc., are surface-level optional and often initialized with `None`.  
- **Suggestion**: Use `Optional[...] = None` directly without `default=` unless there's a specific reason.

#### 5. **Method signatures:**
- Many abstract methods are declared as `def` but are logically intended to be `async`.  
- Some classes have `async def` but may not be used consistently.  
- For example, `BaseLLM.generate_response` is `async`, which is good. But others like `BaseQueue.send_message` are `async` but often the base class declares `def`.  
- **Add `async` exclusively** to methods that inherently are asynchronous for clarity.

#### 6. **Duplicate or redundant code**
- There is some repeated structure, e.g., in `*_models.py`.  
- Some `Optional` fields with default `None` are unnecessarily marked as `Optional[...] = None` and also `default=None`.  
- Cleanup of these for consistency, but it doesn't cause harm.

#### 7. **Design pattern concerns & complexity**
- **Interface segregation** seems adequate, but async method implementations are mixture of `async def` and `def` in abstract classes.  
- Consider consolidating all classes inherit from `ABC` and explicitly `async` where appropriate, to reduce cognitive load.  
- No complex logic or high cyclomatic complexity observed, just adherence issues.

#### 8. **Other minor issues**
- Some classes (e.g., `MediaData`) have `@field_validator` that is not properly annotated (`mode="before"`). The correct syntax in Pydantic v2 should be checked and applied correctly.  
- Minor typo: `arecieves_message` vs. `areceive_message`.
- Consistency in naming: e.g., `context` fields, always use `default=None` for optional.

---

### Score (Out of 100): **75**

- **Strengths**: Clear intention to modernize async patterns, improve typings, and align signatures.
- **Improvements Needed**:
  - Enforce consistent async method declarations.
  - Remove commented/dead code or clarify its purpose.
  - Fix minor type hint issues and redundant optional/default configurations.
  - Enhance interface clarity and documentation.

### Recommendation:  
**Accept with Minor to Moderate revisions**, primarily focusing on asynchronous method consistency, code cleanup, and clear documentation.

---

### References:
- [Python Pydantic v2 Field and Validator docs](https://pydantic.dev/deploy/validation/)
- [Design principles for Async in Python](https://docs.python.org/3/library/abc.html#abc-Attaching-the-ABC-ABC)
- [Clean interface design for async classes](https://medium.com/swlh/async-python-design-patterns-python-asyncio-ec67392a94d6)
